File: Algorithms/124.BinaryTreeMaximumPathSum.py
class Solution:
    # @param root, a tree node
    # @return an integer
    def maxPathSum(self, root):
        self.max_path_sum = None
        self.findMax(root)
        return self.max_path_sum

    def findMax(self, root):
        if not root: return 0
        left = self.findMax(root.left)
        right = self.findMax(root.right)
        self.max_path_sum = max(self.max_path_sum, left + right + root.val)
        ret = root.val + max(left, right)
        return ret if ret > 0 else 0

File: Algorithms/cpp/124.BinaryTreeMaximumPathSum.cpp
class Solution {
public:
    int maxPathSum(TreeNode *root) {
        int max_path_sum = INT_MIN;
        maxPathSum(root, max_path_sum); 
        return max_path_sum;
    }
    int maxPathSum(TreeNode *root, int& max_path_sum) {
        if (root == NULL) return 0;
        int left = maxPathSum(root->left, max_path_sum);
        int right = maxPathSum(root->right, max_path_sum);
        max_path_sum = max(max_path_sum, left + right + root->val);
        int ret = root->val + max(left, right);
        return ret > 0? ret:0;
    }
};
